### PR TITLE
Cluster aware routers for typed

### PR DESCRIPTION
**Pull request by johanandren**
## Purpose
Make cluster routers (and anything that uses the receptionist) aware of unreachability so that nodes that are reachable will be preferred.

## References
Refs #26355 

## Changes
The default listing from the receptionist is changed to be only reachable nodes, a separate method to access both reachable and unreachable is added for special cases.

## Background Context
This will give applications reachability awareness "for free" without changing the main API.

### CODE REVIEW

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala [hunk 1]
```diff
@@ -63,14 +64,34 @@ import scala.concurrent.duration.Deadline
...
+        if (nodes.contains(entryAddress) && !hasTombstone(entry.ref)) {
+          val ref = entry.ref.asInstanceOf[ActorRef[key.Protocol]]
+          all += ref
+          if (!unreachable.contains(entryAddress)) {
+            reachable += ref
+          }
+        }
+        builders
     }
+    (reachable.result(), all.result())
```
Comment (johanandren) [inline akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala]: Mixing mutability and FP like there's no tomorrw. Sorry!

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala [hunk 2]
```diff
@@ -4,11 +4,12 @@
 
 package akka.actor.typed.internal.receptionist
 
+import java.util
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala]: Should we be specific vs an open-ended util?

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 3]
```diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
...
-   * Scala API: Provides a type safe pattern match for listings
+   * Scala API: Provides a type safe pattern match for listings.
+   *
+   * Using it for pattern match like this will return the reachable service instances:
+   *
+   * ```
+   *   case MyServiceKey.Listing(reachable) =>
+   * ```
+   *
+   * In a non-clustered actorsystem this will always be all registered instances
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
   * In a non-clustered `ActorSystem` this will always be all registered instances
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 4]
```diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
...
+   * for a service key. For a clustered environment services on nodes that has
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
   * for a service key. For a clustered environment services on nodes that have
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 5]
```diff
@@ -81,7 +81,19 @@ abstract class ServiceKey[T] extends AbstractServiceKey { key =>
...
+   * been observed unreachable is not among these (note that they could have
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
   * been observed unreachable are not among these (note that they could have
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 6]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     *
+     * In a non-clustered `ActorSystem` this will always be the same as [[#getServiceInstances]].
+     *
+     * For a clustered `ActorSystem` this include both services on nodes that are reachable
+     * and nodes that are unreachable.
+     */
+    def getAllServiceInstances[T](key: ServiceKey[T]): java.util.Set[ActorRef[T]]
+
+    /**
+     * Scala API: `true` only if new services was added or removed and `false` if this listing is
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
     * Scala API: `true` only if new services were added or removed and `false` if this listing is
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 7]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     * only about reachability changes. Useful for subscribers that only cares about [[#allServiceInstances]].
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
     * only about reachability changes. Useful for subscribers only concerned with [[#allServiceInstances]].
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 8]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     *
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def allServiceInstancesChanged: Boolean
+
+    /**
+     * JavaAPI: `true` only if new services was added or removed and `false` if this listing is
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]:
```suggestion
     * JavaAPI: `true` only if new services were added or removed and `false` if this listing is
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 9]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     * only about reachability changes. Useful for subscribers that only cares about [[#getAllServiceInstances]].
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: same as above similar one, or
```suggestion
     * only about reachability changes. Useful for subscribers that only care about [[#getAllServiceInstances]].
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala]: Yeah, some weird autoimport from IntelliJ I think, I'll remove it

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 10]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+    def allServiceInstancesChanged: Boolean
+
+    /**
+     * JavaAPI: `true` only if new services were added or removed and `false` if this listing is
+     * only about reachability changes. Useful for subscribers only concerned with [[#getAllServiceInstances]].
+     *
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def getAllServiceInstancesChanged: Boolean
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: I think normal Java convention is to use `isX` for Boolean

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: I actually think we can have one `isAllServiceInstancesChanged` for both Scala and Java.

For some reason I think `allServiceInstancesChanged` can be misinterpreted as "all instances has changed"

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 11]
```diff
@@ -297,6 +373,20 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+   */
+  def listing[T](
+      key: ServiceKey[T],
+      serviceInstances: java.util.Set[ActorRef[T]],
+      allServiceInstances: java.util.Set[ActorRef[T]],
+      allServiceInstancesChanged: Boolean): Listing =
+    Listing(
+      key,
+      Set[ActorRef[T]](serviceInstances.asScala.toSeq: _*),
+      Set[ActorRef[T]](allServiceInstances.asScala.toSeq: _*),
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: better to use `scala.collection.JavaConverters` ?

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 12]
```diff
@@ -269,15 +281,72 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     * only about reachability changes. Useful for subscribers only concerned with [[#allServiceInstances]].
+     *
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def allServiceInstancesChanged: Boolean
+
+    /**
+     * JavaAPI: `true` only if new services were added or removed and `false` if this listing is
+     * only about reachability changes. Useful for subscribers only concerned with [[#getAllServiceInstances]].
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: , which can ignore all `Listing` messages where this flag is `false`

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala [hunk 13]
```diff
@@ -340,6 +350,50 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
...
+              if (subscribers.nonEmpty) {
+                val (reachable, all) = newRegistry.activeActorRefsFor(serviceKey, selfUniqueAddress)
+                val listing =
+                  ReceptionistMessages.Listing(serviceKey, reachable, all, allServiceInstancesChanged = false)
+                subscribers.foreach(_ ! listing)
+              }
+            }
+          }
+          next(newRegistry)
+
```
Comment (patriknw) [inline akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala]: Should we also add `UnreachableDataCenter`, which would affect all entries owned by members with the dc?
Could be in separate ticket/pr since it's not as important.

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala (hunk 1 above)
Comment (patriknw) [reply akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala]: maybe better to go all in, using only mutable in a `foreach` instead of `foldLeft`?

File: akka-docs/src/main/paradox/typed/routers.md [hunk 14]
```diff
@@ -43,8 +43,8 @@ Java
 The group router is created with a `ServiceKey` and uses the receptionist (see @ref:[Receptionist](actor-discovery.md#receptionist)) to discover
 available actors for that key and routes messages to one of the currently known registered actors for a key.
 
-Since the receptionist is used this means the group router is cluster aware out of the box and will pick up routees
-registered on any node in the cluster (there is currently no logic to avoid routing to unreachable nodes, see [#26355](https://github.com/akka/akka/issues/26355)).
+Since the receptionist is used this means the group router is cluster aware out of the box and will route to 
+registered actors on any node in the cluster that is reachable.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/routers.md]: No changes to the routers in this PR so there must be something missing. I think we said that they would prefer the reachable, but if that is empty it would fall back to all. Better to try to send to someone than just dropping.

**Review by patriknw — COMMENTED** [review]
looking good

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala (hunk 10 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: Perhaps inverting and calling it "onlyReachableChanged" or instead?

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala (hunk 11 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: Wasn't sure of the rationale for the original code being like this but thought there was a reason (tm) so that's the copy-pasta rationale

File: akka-docs/src/main/paradox/typed/routers.md (hunk 14 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/routers.md]: I didn't recall if we were 100% sure we wanted that so didn't go there yet.

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala (hunk 13 above)
Comment (johanandren) [reply akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala]: We'll get both that and an unreachable event per node right? 

Let's ticketify as future improvement.

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala (hunk 10 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: BTW the separate method was mostly to link to the right `allServiceInstances` method from the scaladoc

File: akka-docs/src/main/paradox/typed/routers.md (hunk 14 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/routers.md]: Will make sure this works first then add that on top

File: akka-docs/src/main/paradox/typed/routers.md [hunk 15]
```diff
@@ -88,4 +88,4 @@ it will not give better performance to create more routees than there are thread
 
 Since the router itself is an actor and has a mailbox this means that messages are routed sequentially to the routees
 where it can be processed in parallel (depending on the available threads in the dispatcher).
-In a high throughput use cases the sequential routing could be a bottle neck. Akka Typed does not provide an optimized tool for this.
+In a high throughput use cases the sequential routing could be a bottle neck. Akka Typed does not provide an optimized tool for this.
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/routers.md]: Is there a notion of yet here?

File: akka-docs/src/main/paradox/typed/actor-discovery.md [hunk 16]
```diff
@@ -73,8 +73,15 @@ sends a `Ping` message and when receiving the `Pong` reply it stops.
...
-The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node will eventually reach the same set of actors per `ServiceKey`.
+The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node 
+will eventually reach the same set of actors per `ServiceKey`.
 
-One important difference from a local only receptions is the serialisation concerns, all messages sent to and back from an actor on another node must be serializable, see @ref:[clustering](cluster.md#serialization).
+`Subscription`s and `Find` queries to a clustered receptionist will keep track of cluster reachability and only list 
+registered actors that are reachable. The full set of actors, including unreachable ones, is available through 
+@scala[`Listing.allServiceInstances`]@java[`Listing.getAllServiceInstances`].
+
+One important difference from a local only receptions is the serialisation concerns, all messages sent to and back from
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/actor-discovery.md]:
```suggestion
One important difference from local only receptions are the serialization concerns, all messages sent to and back from 
```

or
```suggestion
One important difference from a local only reception are the serialization concerns, all messages sent to and back from 
```

File: akka-docs/src/main/paradox/typed/routers.md (hunk 15 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/routers.md]: You mean as in "there isn't yet"? We don't plan to introduce one to not end up with the complicated internals we have for untyped to support that.

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 17]
```diff
@@ -269,15 +281,65 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     *
+     * In a non-clustered `ActorSystem` this will always be the same as [[#getServiceInstances]].
+     *
+     * For a clustered `ActorSystem` this include both services on nodes that are reachable
+     * and nodes that are unreachable.
+     */
+    def getAllServiceInstances[T](key: ServiceKey[T]): java.util.Set[ActorRef[T]]
+
+    /**
+     * Returns `false` only if new services were added or removed and `true` if this listing is
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: Should invert this sentence, it reads weird.

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala (hunk 13 above)
Comment (johanandren) [reply akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala]: https://github.com/akka/akka/issues/27327

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala (hunk 10 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: I think it's good that we don't tie this to closely to reachability/unreachability. That way we could have other aspects for choosing preferred vs all in the future. E.g. other health or load metrics.

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 18]
```diff
@@ -269,15 +281,65 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+
+    /**
+     * Returns `true` if this listing is only about reachability changes and `false` if new services were
+     * actually added or removed to the receptionist. Useful for subscribers only concerned with [[#allServiceInstances]]
+     * or [[#getAllServiceInstances]] that can then ignore `Listing`s related to reachability.
+     *
+     * In a non-clustered `ActorSystem` this will be `false` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `false`.
+     */
+    def onlyReachabilityChanged: Boolean
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: I'm not sure this is a good name change. Nowhere else in the API signatures we have reachability. Could be good to not tie it too close to unreachable/reachable so that we can use it for preferred service instances based on other other health or load metrics in the future.

```
def isAllServiceInstancesChanged: Boolean
```

File: akka-docs/src/main/paradox/typed/actor-discovery.md [hunk 19]
```diff
@@ -73,8 +73,15 @@ sends a `Ping` message and when receiving the `Pong` reply it stops.
 
 ## Cluster Receptionist
 
-The `Receptionist` also works in a cluster, an actor registered to the receptionist will appear in the receptionist of the other nodes of the cluster.
+The `Receptionist` also works in a cluster, an actor registered to the receptionist will appear in the receptionist 
+of the other nodes of the cluster.
 
-The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node will eventually reach the same set of actors per `ServiceKey`.
+The state for the receptionist is propagated via @ref:[distributed data](../distributed-data.md) which means that each node
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/actor-discovery.md]: Distributed Data

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala (hunk 18 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: I think `isAllServiceInstancesChanged` is hard to parse though. What about `wasRegisteredActorsChanged`?

Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: `wasServicesAddedOrRemoved`

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: that is a good name!

**Review by helena — APPROVED** [review]
Looks great

File: akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala [hunk 20]
```diff
@@ -340,6 +351,50 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
...
+                val listing =
+                  ReceptionistMessages.Listing(serviceKey, reachable, all, servicesWereAddedOrRemoved = true)
+                subscribers.foreach(_ ! listing)
+              }
+            }
+          }
+          next(newRegistry)
+
+        case NodeReachable(uniqueAddress) =>
+          val keysForNode = registry.keysFor(uniqueAddress)
```
Comment (chbatey) [inline akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala]: maybe extract method for this? Only difference is the log line

File: akka-cluster-typed/src/multi-jvm/scala/akka/cluster/typed/internal/ClusterReceptionistUnreachabilitySpec.scala [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+      typedSystem.receptionist ! Receptionist.Subscribe(MyServiceKey, receptionistProbe.ref)
+      val listing = receptionistProbe.expectMessageType[Receptionist.Listing]
+      listing.serviceInstances(MyServiceKey) should ===(Set.empty)
+      listing.allServiceInstances(MyServiceKey) should ===(Set.empty)
+      listing.servicesWereAddedOrRemoved should ===(true)
+      enterBarrier("all subscribed")
+    }
+
+    "form a cluster" in {
+      runOn(first) {
```
Comment (chbatey) [inline akka-cluster-typed/src/multi-jvm/scala/akka/cluster/typed/internal/ClusterReceptionistUnreachabilitySpec.scala]: Use MultiNodeTypedClusterSpec.formCluster

**Review by chbatey — APPROVED** [review]
LGTM with two comments

File: akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala [hunk 22]
```diff
@@ -269,15 +281,68 @@ object Receptionist extends ExtensionId[Receptionist] {
...
+     * When `false` the event is only about reachability changes - meaning that the full set of actors
+     * ([[#allServiceInstances]] or [[#getAllServiceInstances]]) is the same as the previous `Listing`.
+     *
+     * knowing this is useful for subscribers only concerned with [[#allServiceInstances]] or [[#getAllServiceInstances]]
+     * that can then ignore `Listing`s related to reachability.
+     *
+     * In a non-clustered `ActorSystem` this will be `true` for all listings.
+     * For `Find` queries and the initial listing for a `Subscribe` this will always be `true`.
+     */
+    def servicesWereAddedOrRemoved: Boolean
```
Comment (raboof) [inline akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala]: :+1:

**Review by raboof — APPROVED** [review]
